crates/bevy_pbr/src/light_probe/generate.rs

```diff
@@ -0,0 +1,1084 @@
+        }
+
+        // Create irradiance bind group
+        let irradiance_constants = FilteringConstants {
+            mip_level: 0.0,
+            // 32 phi, 32 theta = 1024 samples total
+            sample_count: 1024,
+            roughness: 1.0,
+            blue_noise_size: texture_size,
+        };
+
+        let mut irradiance_constants_buffer = UniformBuffer::from(irradiance_constants);
+        irradiance_constants_buffer.write_buffer(&render_device, &queue);
+
+        // create a 2d array view
+        let irradiance_map =
+            env_map_light
+                .diffuse_map
+                .texture
+                .create_view(&TextureViewDescriptor {
+                    dimension: Some(TextureViewDimension::D2Array),
+                    ..Default::default()
+                });
+
+        let irradiance_bind_group = render_device.create_bind_group(
+            "irradiance_bind_group",
+            &layouts.irradiance,
+            &BindGroupEntries::with_indices((
+                (0, &textures.environment_map.default_view),
+                (1, &samplers.linear),
+                (2, &irradiance_map),
+                (3, &irradiance_constants_buffer),
+                (4, &stbn_texture.texture_view),
+            )),
+        );
+
+        // Create copy bind group (source env map → destination mip0)
+        let src_view = env_map_light
+            .environment_map
+            .texture
+            .create_view(&TextureViewDescriptor {
+                dimension: Some(TextureViewDimension::D2Array),
+                ..Default::default()
+            });
+
+        let dst_view = create_storage_view(&textures.environment_map.texture, 0, &render_device);
+
+        let copy_bind_group = render_device.create_bind_group(
+            "copy_mip0_bind_group",
+            &layouts.copy,
+            &BindGroupEntries::with_indices(((0, &src_view), (1, &dst_view))),
+        );
+
+        commands.entity(entity).insert(GeneratorBindGroups {
+            spd: spd_bind_group,
+            radiance: radiance_bind_groups,
+            irradiance: irradiance_bind_group,
+            copy: copy_bind_group,
+        });
+    }
+}
+
+/// Helper function to create a storage texture view for a specific mip level
+fn create_storage_view(texture: &Texture, mip: u32, _render_device: &RenderDevice) -> TextureView {
+    texture.create_view(&TextureViewDescriptor {
+        label: Some(format!("storage_view_mip_{mip}").as_str()),
+        format: Some(texture.format()),
+        dimension: Some(TextureViewDimension::D2Array),
+        aspect: TextureAspect::All,
+        base_mip_level: mip,
+        mip_level_count: Some(1),
+        base_array_layer: 0,
+        array_layer_count: Some(texture.depth_or_array_layers()),
+        usage: Some(TextureUsages::STORAGE_BINDING),
+    })
+}
+
+/// SPD Node implementation that handles both parts of the downsampling (mips 0-12)
+pub struct SpdNode {
```

Comment:
Imo, try to split up this file somewhat. I think you should make a new module for all this stuff and the filtering shader and noise. 

Node is a good candidate for going in a separate file.

Maybe copy the structure I did for solari/realtime.

Reply:
Splitting up this file is a good suggestion, but I think it is out of scope for this PR and could be addressed as a follow-up. Unfortunately I could not allow as good of a modularization as Solari without writing some more boilerplate code and having too many files under `light_probe` folder. We might consider creating separate folders at the higher level to keep the directory structure flat.